granary/reddit.py

```diff
@@ -0,0 +1,218 @@
+      'id': self.tag_uri(id),
+      'objectType': 'note',
+      'published': published,
+      'to': [],
+      }
+
+    user = thing.author
+    if user:
+      obj['author'] = self.user_to_actor(user)
+      username = obj['author'].get('username')
+
+    obj['to'].append({
+      'objectType': 'group',
+      'alias': '@public',
+      })
+
+    obj['url'] = self.BASE_URL + thing.permalink
+
+    if type == 'submission':
+      obj['content'] = thing.title
+    elif type == 'comment':
+      obj['content'] = thing.body
+
+    return self.postprocess_object(obj)
+
+
+  def praw_to_activity(self, thing, type):
+    """Converts a praw submission or comment to an activity.
+    https://praw.readthedocs.io/en/latest/code_overview/models/submission.html
+    https://praw.readthedocs.io/en/latest/code_overview/models/comment.html
+
+    Args:
+      thng: a praw object, Sumbission or Comment
+      type: string to denote whether to get submission or comment content
+
+    Returns:
+      an ActivityStreams activity dict, ready to be JSON-encoded
+    """
+    obj = self.praw_to_object(thing, type)
+    actor = obj['author']
+
+    activity = {
+      'verb': 'post',
+      'id': thing.id,
+      'url': self.BASE_URL + thing.permalink,
+      'actor': actor,
+      'object': obj
+      }
+
+    return self.postprocess_activity(activity)
+
+  def fetch_replies(self, r, activities):
+    """Fetches and injects reddit comments into a list of activities, in place.
+
+    limitations: Only includes top level comments
+    Args:
+      r: praw api object for querying submissions in activities
+      activities: list of activity dicts
+
+    Returns:
+      same activities list
+    """
+    for activity in activities:
+      subm = r.submission(id=activity['id'])
+
+      # for v0 we will use just the top level comments because threading is hard
+      subm.comments.replace_more()
+      replies = []
+      for top_level_comment in subm.comments:
+        replies.append(self.praw_to_activity(top_level_comment, 'comment'))
+
+      items = [r['object'] for r in replies]
+      activity['object']['replies'] = {
+        'items': items,
+        'totalItems': len(items),
+        }
+
+  def get_activities_response(self, user_id=None, group_id=None, app_id=None,
+                activity_id=None, start_index=0, count=0,
+                etag=None, min_id=None, cache=None,
+                fetch_replies=False, fetch_likes=False,
+                fetch_shares=False, fetch_events=False,
+                fetch_mentions=False, search_query=None, **kwargs):
+    """
+    Fetches reddit submissions and ActivityStreams activities.
+
+    Currently only implements activity_id, search_query and fetch_replies
+    """
+
+    activities = []
+
+    r = get_reddit_api(self.refresh_token)
+    r.read_only = True
+
+    if activity_id:
+      subm = r.submission(id=activity_id)
+      activities.append(self.praw_to_activity(subm, 'submission'))
+
+    if search_query:
```

Comment:
these are generally exclusive, ie either we fetch a specific activity id, or we search, or etc... example:

https://github.com/snarfed/granary/blob/6e97787c72a139709e011587a65722ec64212fe4/granary/twitter.py#L259-L292

Reply:
I noticed that in twitter.py, but I figured it would be fine to let them be used simultaneously if someone really wanted to in this case.